File: app-service-comments/src/main/java/com/github/arhor/aws/graphql/federation/comments/config/ConfigureCaching.java
package com.github.arhor.aws.graphql.federation.comments.config;

import com.github.arhor.aws.graphql.federation.comments.util.Caches;
import com.github.benmanes.caffeine.cache.Caffeine;
import org.springframework.boot.autoconfigure.cache.CacheManagerCustomizer;
import org.springframework.cache.annotation.EnableCaching;
import org.springframework.cache.caffeine.CaffeineCacheManager;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import java.time.Duration;

@EnableCaching
@Configuration(proxyBeanMethods = false)
public class ConfigureCaching {

    @Bean
    public CacheManagerCustomizer<CaffeineCacheManager> caffeineCacheManagerCustomizer() {
        return (it) -> it.registerCustomCache(
            Caches.IDEMPOTENT_ID_SET.name(),
            Caffeine.newBuilder()
                .maximumSize(10_000)
                .expireAfterAccess(Duration.ofMinutes(30))
                .build()
        );
    }
}

File: app-service-comments/src/main/java/com/github/arhor/aws/graphql/federation/comments/config/ConfigureCaching.kt
package com.github.arhor.aws.graphql.federation.comments.config

import com.github.benmanes.caffeine.cache.Caffeine
import graphql.ExecutionInput
import graphql.execution.preparsed.PreparsedDocumentEntry
import graphql.execution.preparsed.PreparsedDocumentProvider
import org.springframework.cache.annotation.EnableCaching
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import java.time.Duration
import java.util.concurrent.Executor
import java.util.function.Function

private typealias ParsingValidator = Function<ExecutionInput, PreparsedDocumentEntry>

@EnableCaching
@Configuration(proxyBeanMethods = false)
class ConfigureCaching {

    @Bean
    @Suppress("OVERRIDE_DEPRECATION")
    fun asyncCachePreParsedDocumentProvider(executor: Executor): PreparsedDocumentProvider =
        object : PreparsedDocumentProvider {
            private val cache = buildAsyncCache<String, PreparsedDocumentEntry> {
                maximumSize(250)
                expireAfterAccess(Duration.ofMinutes(10))
                executor(executor)
            }

            override fun getDocument(execInput: ExecutionInput, validator: ParsingValidator) =
                cache.synchronous().get(execInput.query) { _ -> validator.apply(execInput) }

            override fun getDocumentAsync(execInput: ExecutionInput, validator: ParsingValidator) =
                cache.get(execInput.query) { _ -> validator.apply(execInput) }
        }

    private inline fun <K, V> buildAsyncCache(init: Caffeine<Any, Any>.() -> Unit) =
        Caffeine.newBuilder().apply { init() }.buildAsync<K, V>()
}
